File: leetcode/2894.cpp
class Solution {
  public:
    int differenceOfSums(int n, int m) {
        auto recur = [m](this auto &&self, int n, int non = 0,
                         int div = 0) -> int {
            if (n == 0)
                return non - div;

            return (n % m == 0) ? self(n - 1, non, div + n)
                                : self(n - 1, non + n, div);
        };

        return recur(n);
    }
};

File: leetcode/2894.rs
impl Solution {
    pub fn difference_of_sums(n: i32, m: i32) -> i32 {
        fn recur(n: i32, non: i32, div: i32, m: i32) -> i32 {
            if n == 0 {
                return non - div;
            }

            if n % m == 0 {
                return recur(n - 1, non, div + n, m);
            }

            return recur(n - 1, non + n, div, m);
        }

        recur(n, 0, 0, m)
    }
}
